contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,388 @@
+        //get market
+        IOptionMarket market = IOptionMarket(token.market());
+
+        if (block.timestamp < market.expiryTime()) {
+            return false;
+        }
+        if (exerciseProfit(options, optionID, amount) > 0) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @notice Purchases the options contract.
+     * @param optionTerms is the terms of the option contract
+     * @param amount is the purchase amount in Wad units (10**18)
+     * @param maxCost is the max amount of paymentToken to be paid for the option (to avoid sandwich attacks, ...)
+     */
+    function purchase(
+        ProtocolAdapterTypes.OptionTerms calldata optionTerms,
+        uint256 amount,
+        uint256 maxCost
+    ) external payable override returns (uint256) {
+      require(
+          block.timestamp < optionTerms.expiry,
+          "Cannot purchase after expiry"
+      );
+
+      //get token address
+      address tokenAddress = lookupOToken(optionTerms);
+      if(tokenAddress == address(0)){
+        populateOTokenMappings(optionTerms);
+      }
+
+      tokenAddress = lookupOToken(optionTerms);
+
+      require(tokenAddress != address(0), "Market needs to exist!");
+
+      // Get strike index, whether is long
+      OptionType memory optionType = addressToOptionType[tokenAddress];
+      //get token
+      IOptionToken token = IOptionToken(tokenAddress);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+      IERC20 baseToken = market.baseToken();
+
+      uint256 balanceBefore = baseToken.uniBalanceOf(address(this));
+      baseToken.uniTransferFromSenderToThis(amount);
+      uint256 balanceAfter = baseToken.uniBalanceOf(address(this));
+
+      uint256 amountIn;
+
+      if(baseToken.isETH()){
+        amountIn = market.buy{value: address(this).balance}(
+          optionType.isLongToken,
+          optionType.strikeIndex,
+          amount,
+          maxCost
+        );
+      }else{
+        baseToken.safeApprove(address(market), balanceAfter.sub(balanceBefore));
+        amountIn = market.buy(
+          optionType.isLongToken,
+          optionType.strikeIndex,
+          amount,
+          maxCost
+        );
+      }
+
+      emit Purchased(
+          msg.sender,
+          _name,
+          optionTerms.underlying,
+          amountIn,
+          0
+      );
+    }
+
+    /**
+     * @notice Exercises the options contract.
+     * @param options is the address of the options contract
+     * @param optionID is the ID of the option position in non fungible protocols like Hegic.
+     * @param amount is the amount of tokens or options contract to exercise. Only relevant for fungle protocols like Opyn
+     * @param recipient is the account that receives the exercised profits. This is needed since the adapter holds all the positions and the msg.sender is an instrument contract.
+     */
+    function exercise(
```

Comment:
Should exercise() be only for exercising after market expiry / settlement OR can it also be for selling tokens before expiry?

Reply:
Let's limit the scope to just exercising after market expiry, we can add another ProtocolAdapter function to sell options in the future